# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv('e_commerce.csv')
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

数据的评估分为结构和内容两部分  
数据结构，即整齐度，数据是否符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准；  
数据内容，即干净度，数据是否存在缺失值，重复值，不一致值以及无效值等问题。

### 评估数据整齐度

In [3]:
original_data.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
304222    563565     22467                    GUMBALL COAT RACK         2   
469282    576592     21200  MULTICOLOUR HONEYCOMB PAPER GARLAND         2   
50887     540566     22224               WHITE LOVEBIRD LANTERN         2   
246247    558684     85053           FRENCH ENAMEL CANDLEHOLDER        12   
118320    546428     22320          BIRDS MOBILE VINTAGE DESIGN         3   
398766    571257     22416   SET OF 36 DOILIES SPACEBOY DESIGN          1   
436525    574169     23344            JUMBO BAG 50'S CHRISTMAS         30   
366728    568791     21876                        POTTERING MUG        12   
7453      537041     85054              FRENCH ENAMEL POT W LID         2   
287560    562106     21205         MULTICOLOUR 3D BALLS GARLAND         1   

             InvoiceDate  UnitPrice  CustomerID         Country  
304222   8/17/2011 14:22       2.55     17338.0  United Kingdom  
469282  11/15/2011 14:47       1.65     13614.0  United Kingdom  
50887    1/10/2011 10:58       2.95     17811.0  United Kingdom  
246247    7/1/2011 11:29       2.10     14016.0            EIRE  
118320   3/13/2011 14:55       5.95     12681.0          France  
398766  10/14/2011 17:17       2.92         NaN  United Kingdom  
436525   11/3/2011 11:40       2.08     17491.0  United Kingdom  
366728    9/29/2011 9:51       1.65     16650.0  United Kingdom  
7453     12/5/2010 10:36       2.95     16033.0  United Kingdom  
287560    8/2/2011 15:19       2.55     14076.0  United Kingdom

数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，无需对数据整齐度进行调整。

### 评估数据干净度

In [4]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


`Description`产品名称和`CustomerID`顾客id存在缺失值，下面具体看一下缺失值是否影响分析目标；  
此外，`InvoiceDate`的数据类型应为datetime日期，`CustomerID`数据类型为object字符串,应该进行数据类型转换。

#### 缺失数据

`Description`产品名称缺失1454行数据。

In [5]:
original_data[original_data['Description'].isnull()].sample(10)

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
202328    554377     79164         NaN        47  5/24/2011 11:34        0.0   
218104    556015     37501         NaN       -13   6/8/2011 11:50        0.0   
280746    561490     21275         NaN         3  7/27/2011 13:54        0.0   
19999     538001     35958         NaN        10  12/9/2010 11:45        0.0   
319910    564929     37501         NaN         1  8/31/2011 12:46        0.0   
96679     544579     21651         NaN        -6  2/21/2011 14:05        0.0   
281569    561538    84927D         NaN        -1   7/28/2011 9:58        0.0   
42500     539996     21784         NaN       -10   1/4/2011 10:12        0.0   
304304    563568     20774         NaN        12  8/17/2011 14:45        0.0   
78424     542879     22242         NaN       -19   2/1/2011 12:52        0.0   

        CustomerID         Country  
202328         NaN  United Kingdom  
218104         NaN  United Kingdom  
280746         NaN  United Kingdom  
19999          NaN  United Kingdom  
319910         NaN  United Kingdom  
96679          NaN  United Kingdom  
281569         NaN  United Kingdom  
42500          NaN  United Kingdom  
304304         NaN  United Kingdom  
78424          NaN  United Kingdom

随机取10行`Description`缺失的数据观察得出假设，当`Description`为Nan时，`Unitprice`均为0，下面验证猜想。

In [6]:
original_data[(original_data['Description'].isnull()) & (original_data['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

猜想正确，筛选出来结果数量为0条，说明缺失`Description`值的数据，同时也不具备有效的`UnitPrice`值。

`Description`表示产品名称，`UnitPrice`表示产品单价，都是进行后续商品交易分析的重要变量。如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此这些后续可以被删除。

`CustomerID`顾客id缺失135080行数据。

In [7]:
original_data[original_data['CustomerID'].isnull()].sample(10)

InvoiceNo StockCode                        Description  Quantity  \
41817     539958    84968A     SET OF 16 VINTAGE ROSE CUTLERY         9   
523685    580519     21484        CHICK GREY HOT WATER BOTTLE         3   
60049     541421     21731      RED TOADSTOOL LED NIGHT LIGHT         1   
212673    555502     22799         SWEETHEART WIRE FRUIT BOWL         1   
160315    550458     22685      FRENCH BLUE METAL DOOR SIGN 0         1   
290943    562420     22620        4 TRADITIONAL SPINNING TOPS         1   
184201    552677     21033    JUMBO BAG CHARLIE AND LOLA TOYS         2   
478113    577078    84596F       SMALL MARSHMALLOWS PINK BOWL         2   
53663     540848     20663  QUEEN OF THE SKIES HOLIDAY PURSE          1   
57136     541132     21253     SET OF PICTURE FRAME  STICKERS         1   

             InvoiceDate  UnitPrice  CustomerID         Country  
41817   12/23/2010 13:26       8.47         NaN  United Kingdom  
523685   12/4/2011 14:51       9.13         NaN  United Kingdom  
60049    1/17/2011 17:44       3.29         NaN  United Kingdom  
212673    6/3/2011 15:37       8.50         NaN  United Kingdom  
160315   4/18/2011 13:13       2.46         NaN  United Kingdom  
290943    8/4/2011 16:38       2.46         NaN  United Kingdom  
184201   5/10/2011 15:07       4.13         NaN  United Kingdom  
478113  11/17/2011 15:17       0.83         NaN  United Kingdom  
53663     1/12/2011 9:26       2.51         NaN  United Kingdom  
57136    1/14/2011 10:17       5.79         NaN  United Kingdom

In [8]:
original_data[(original_data['Description'].isnull()) & (original_data['CustomerID'].isnull())]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

In [9]:
original_data[(original_data['CustomerID'].isnull()) & (original_data.duplicated())].info()

<class 'pandas.core.frame.DataFrame'>
Index: 43 entries, 14359 to 514741
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   InvoiceNo    43 non-null     object 
 1   StockCode    43 non-null     object 
 2   Description  43 non-null     object 
 3   Quantity     43 non-null     int64  
 4   InvoiceDate  43 non-null     object 
 5   UnitPrice    43 non-null     float64
 6   CustomerID   0 non-null      float64
 7   Country      43 non-null     object 
dtypes: float64(2), int64(1), object(5)
memory usage: 3.0+ KB


`Description`和`CustomerID`同时为Nan的数据有1454行，执行删除`Description`缺失数据时，均会被删除，无需再次考虑。  
存在`CustomerID`为Nan时，完全重复的数据，此处暂不考虑，放到*评估重复数据*小节讨论。

`CustomerID`表示客户编号，不是分析畅销商品的必要变量。并且从输出结果来看，有些`CustomerID`缺失的交易数据仍然有效，因此保留此变量为空的观察值。

#### 重复数据

In [10]:
original_data[original_data.duplicated()]

InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
541675    581538     22068        BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992             REVOLVER WOODEN RULER          1   
541699    581538     22694                       WICKER STAR          1   
541701    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

            InvoiceDate  UnitPrice  CustomerID         Country  
517     12/1/2010 11:45       1.25     17908.0  United Kingdom  
527     12/1/2010 11:45       2.10     17908.0  United Kingdom  
537     12/1/2010 11:45       2.95     17908.0  United Kingdom  
539     12/1/2010 11:45       4.95     17908.0  United Kingdom  
555     12/1/2010 11:49       2.95     17920.0  United Kingdom  
...                 ...        ...         ...             ...  
541675  12/9/2011 11:34       0.39     14446.0  United Kingdom  
541689  12/9/2011 11:34       2.49     14446.0  United Kingdom  
541692  12/9/2011 11:34       1.95     14446.0  United Kingdom  
541699  12/9/2011 11:34       2.10     14446.0  United Kingdom  
541701  12/9/2011 11:34       2.08     14446.0  United Kingdom  

[5268 rows x 8 columns]

根据数据变量的含义来看，虽然`InvoiceNo`、`StockCode`和`CustomerID`都是唯一标识符，但一次交易可能包含多件商品，因此`InvoiceNo`可以存在重复，不同交易可以包含同件商品，因此`StockCode`可以存在重复，顾客可以进行多次交易或下单多个商品，因此`CustomerID`也可以存在重复。但是当`InvoiceNo` `InvoiceDate` `CustomerID` `StockCode` `Quantity`同时重复的情况不应当出现，即一个用户在同一时间，同一张发票上购买了同一产品，但是被分开记录了，所以该情况可能是记录人员重复登录销售信息，这些数据会影响我们分析目的的实现，不应予以考虑，应当删除重复记录。  

`CustomerID`为Nan时，完全重复的数据，仅有43行，包含在所有重复数据里一同删除。


#### 不一致数据

不一致数据可能存在于`Country`变量中，我们要查看是否存在多个不同值指代同一国家的情况。

In [11]:
original_data['Country'].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


从`Country`变量值来看，`"USA"`、`"United States"`均在表示美国，`"United Kingdom"`、`"UK"`、`"U.K."`均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

#### 失效/无效数据

In [12]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，`Quantity`和`UnitPrice`存在负数，会对后续数值分析产生影响。

因此，我们先筛选出`Quantity`数值为负数的观察值，进一步评估其含义。

In [13]:
original_data[(original_data['Quantity'] < 0)]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，`Quantity`变量为负数的观察值，`InvoiceNo`似乎均以C开头，表示订单被取消。

但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。

In [14]:
original_data[(original_data['Quantity'] < 0) & (~ original_data['InvoiceNo'].str.contains('C'))]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。

但根据以上输出结果，这些筛选出的观察值的`UnitPrice`观察值均为0，因此增加`UnitPrice`的条件进行验证。

In [15]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当`Quantity`变量为负数的时候，观察值满足以下条件之一：
- `InvoiceNo`以C开头，表示订单被取消
- `UnitPrice`为0，说明单价为0英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将`Quantity`为负数的观察值删除。

接下来筛选出`UnitPrice`为零或
负数的观察值。

In [16]:
original_data[original_data['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，`UnitPrice`为负数的观察值都是坏账调账，不属于实际商品交易数据，因此也在数据清理步骤中也将其删除。

In [17]:
original_data[original_data['UnitPrice'] == 0]

InvoiceNo StockCode                   Description  Quantity  \
622       536414     22139                           NaN        56   
1970      536545     21134                           NaN         1   
1971      536546     22145                           NaN         1   
1972      536547     37509                           NaN         1   
1987      536549    85226A                           NaN         1   
...          ...       ...                           ...       ...   
536981    581234     72817                           NaN        27   
538504    581406    46000M  POLYESTER FILLER PAD 45x45cm       240   
538505    581406    46000S  POLYESTER FILLER PAD 40x40cm       300   
538554    581408     85175                           NaN        20   
538919    581422     23169                       smashed      -235   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52        0.0         NaN  United Kingdom  
1970    12/1/2010 14:32        0.0         NaN  United Kingdom  
1971    12/1/2010 14:33        0.0         NaN  United Kingdom  
1972    12/1/2010 14:33        0.0         NaN  United Kingdom  
1987    12/1/2010 14:34        0.0         NaN  United Kingdom  
...                 ...        ...         ...             ...  
536981  12/8/2011 10:33        0.0         NaN  United Kingdom  
538504  12/8/2011 13:58        0.0         NaN  United Kingdom  
538505  12/8/2011 13:58        0.0         NaN  United Kingdom  
538554  12/8/2011 14:06        0.0         NaN  United Kingdom  
538919  12/8/2011 15:24        0.0         NaN  United Kingdom  

[2515 rows x 8 columns]

在*缺失数据*小节，我们验证了当`Description`为Nan值时，`UnitPrice`也均为0.0，此数据有1454行。  

在本小节，我们验证了当`Quantity`< 0 时，且`InvoiceNo`不以C开头，`UnitPrice`为0.0的数据有1336行。  

我们需要筛选出来除以上条件外的其他`UnitPrice` 为0.0 的观察值。

In [18]:
original_data[(original_data['UnitPrice'] == 0) & (original_data['Description'].notnull()) & (original_data["Quantity"] > 0)]

InvoiceNo StockCode                       Description  Quantity  \
6391      536941     22734                            amazon        20   
6392      536942     22139                            amazon        15   
9302      537197     22841      ROUND CAKE TIN VINTAGE GREEN         1   
14335     537534     85064      CREAM SWEETHEART LETTER RACK         1   
14336     537534     84832  ZINC WILLIE WINKIE  CANDLE STICK         1   
...          ...       ...                               ...       ...   
524622    580609     22927                            Amazon         1   
535325    581202     23404                             check        41   
535334    581211     22142                             check        14   
538504    581406    46000M      POLYESTER FILLER PAD 45x45cm       240   
538505    581406    46000S      POLYESTER FILLER PAD 40x40cm       300   

            InvoiceDate  UnitPrice  CustomerID         Country  
6391    12/3/2010 12:08        0.0         NaN  United Kingdom  
6392    12/3/2010 12:08        0.0         NaN  United Kingdom  
9302    12/5/2010 14:02        0.0     12647.0         Germany  
14335   12/7/2010 11:48        0.0         NaN  United Kingdom  
14336   12/7/2010 11:48        0.0         NaN  United Kingdom  
...                 ...        ...         ...             ...  
524622  12/5/2011 11:41        0.0         NaN  United Kingdom  
535325  12/7/2011 18:30        0.0         NaN  United Kingdom  
535334  12/7/2011 18:36        0.0         NaN  United Kingdom  
538504  12/8/2011 13:58        0.0         NaN  United Kingdom  
538505  12/8/2011 13:58        0.0         NaN  United Kingdom  

[587 rows x 8 columns]

在挖掘畅销产品的过程中，销售单价为零的观察值可能会有一定的意义，但也可能是数据中的异常情况。下面是一些考虑销售单价为零观察值意义的因素：

- 免费赠品或促销活动：销售单价为零的观察值可能代表客户获得了免费赠品或者参与了某种促销活动。这种情况下，这些观察值可能有助于理解营销策略的有效性以及客户的购买行为。

- 数据质量问题：销售单价为零的观察值也可能是由于数据质量问题导致的，例如数据输入错误或者数据缺失。在这种情况下，这些观察值可能会干扰畅销产品的挖掘分析，需要谨慎处理或者排除。

- 异常销售情况：销售单价为零的观察值可能表示了某些异常的销售情况，例如退货、取消订单或者系统错误等。这些情况下，这些观察值可能不代表真正的销售情况，需要谨慎处理。

## 清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括：  
- 删除完全重复5268行数据；
- 删除缺失`Description`值的数据；
- 删除`Quantity`为负数的观察值；
- 删除`UnitPrice`为负数的观察值（都是坏账调整）；
- 统一`Country`变量值，`"United States"`转换为`"USA"`，`"United Kingdom"`、`"U.K."`转换为`"UK"`；
- 转换`InvoiceDate`的数据类型为datetime日期，转换`CustomerID`数据类型为object字符串；

为了区分开经过清理的数据和原始的数据，我们创建新的变量`cleaned_data`，让它为`original_data`复制出的副本。我们之后的清理步骤都将被运用在`cleaned_data`上。

In [19]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

- 删除完全重复5268行数据

In [20]:
cleaned_data[cleaned_data.duplicated()]

InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
541675    581538     22068        BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992             REVOLVER WOODEN RULER          1   
541699    581538     22694                       WICKER STAR          1   
541701    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

            InvoiceDate  UnitPrice  CustomerID         Country  
517     12/1/2010 11:45       1.25     17908.0  United Kingdom  
527     12/1/2010 11:45       2.10     17908.0  United Kingdom  
537     12/1/2010 11:45       2.95     17908.0  United Kingdom  
539     12/1/2010 11:45       4.95     17908.0  United Kingdom  
555     12/1/2010 11:49       2.95     17920.0  United Kingdom  
...                 ...        ...         ...             ...  
541675  12/9/2011 11:34       0.39     14446.0  United Kingdom  
541689  12/9/2011 11:34       2.49     14446.0  United Kingdom  
541692  12/9/2011 11:34       1.95     14446.0  United Kingdom  
541699  12/9/2011 11:34       2.10     14446.0  United Kingdom  
541701  12/9/2011 11:34       2.08     14446.0  United Kingdom  

[5268 rows x 8 columns]

In [21]:
cleaned_data.drop_duplicates(inplace = True)
cleaned_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[536641 rows x 8 columns]

检查是否还存在完全重复数据：

In [22]:
541909-536641

5268

In [23]:
cleaned_data.duplicated().sum()

0

- 删除缺失`Description`值的数据

In [24]:
cleaned_data.dropna(subset = 'Description' , inplace = True)

检查是否还存在`Description`缺失的数据：

In [25]:
536641-535187

1454

In [26]:
cleaned_data["Description"].isnull().sum()

0

- 删除`Quantity`为负数的观察值

In [27]:
cleaned_data = cleaned_data[cleaned_data['Quantity'] >= 0]

检查是否还存在`Quantity`为负的数据：

In [28]:
cleaned_data[cleaned_data['Quantity'] < 0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [29]:
len(cleaned_data[cleaned_data['Quantity'] < 0])

0

- 删除`UnitPrice`为负数的观察值（都是坏账调整）

In [30]:
cleaned_data = cleaned_data[cleaned_data['UnitPrice'] >= 0]

检查是否还存在`UnitPrice`为负的数据：

In [31]:
len(cleaned_data[cleaned_data['UnitPrice'] < 0])

0

- 统一`Country`变量值，`"USA"`替换为`"United States"`，`"UK"`、`"U.K."`替换为`"United Kingdom"`

In [32]:
cleaned_data = cleaned_data.replace({'USA':'United States' , 'UK':'United Kingdom' , 'U.K.':'United Kingdom'})

检查是否还存在'USA','UK','U.K.':

In [33]:
len(cleaned_data[(cleaned_data['Country'] == 'USA') | (cleaned_data['Country'] == 'UK') | (cleaned_data['Country'] == 'U.K.')])

0

- 转换`InvoiceDate`的数据类型为datetime日期，转换`CustomerID`数据类型为object字符串

In [34]:
cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data['InvoiceDate'])
cleaned_data['InvoiceDate'] 

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 525460, dtype: datetime64[ns]

In [35]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].astype(str)
cleaned_data['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 525460, dtype: object

In [36]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].str.slice(0,-2)
cleaned_data['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 525460, dtype: object

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为`e_commerce_cleaned(2.0).csv`。

In [38]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [39]:
cleaned_data.to_csv("e_commerce_cleaned(2.0).csv", index=False)

In [42]:
pd.read_csv("e_commerce_cleaned(2.0).csv").head()

D:\TEMP 变量缓存\ipykernel_23024\4185021829.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned(2.0).csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom